### ISSUE TITLE
Submission: Application Skeleton: Generating Synthetic Applications for Infrastructure Research

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#17]
**Submitting author:** @zhaozhang (<a href="http://orcid.org/0000-0001-5921-0035">Zhao Zhang</a>)
**Repository:** <a href="https://github.com/applicationskeleton/Skeleton" target ="_blank">https://github.com/applicationskeleton/Skeleton</a>
**Version:** v1.2
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.13750" target ="_blank">http://dx.doi.org/10.5281/zenodo.13750</a>

## Status

[![status](http://joss.theoj.org/papers/6b34fdbf8215c1ecd4941afc8be7a7e6/status.svg)](http://joss.theoj.org/papers/6b34fdbf8215c1ecd4941afc8be7a7e6)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/6b34fdbf8215c1ecd4941afc8be7a7e6"><img src="http://joss.theoj.org/papers/6b34fdbf8215c1ecd4941afc8be7a7e6/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/6b34fdbf8215c1ecd4941afc8be7a7e6/status.svg)]http://joss.theoj.org/papers/6b34fdbf8215c1ecd4941afc8be7a7e6
```

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/applicationskeleton/Skeleton">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.2)?
- [x] **Archive:** Does the <a target="_blank" href="http://dx.doi.org/10.5281/zenodo.13750">software archive</a> resolve?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** Have the performance claims of the software been confirmed?

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the functionality of the software documented to a satisfactory level (e.g. API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

Compiled paper PDF: [paper.pdf](https://github.com/openjournals/joss-reviews/files/259667/paper.pdf)

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g. papers, datasets, software)?

**Comment by arfon** [comment]
/ cc @openjournals/joss-reviewers - would anyone be willing to review this submission?

If you would like to review this submission then please comment on this thread so that others know you're doing a review (so as not to duplicate effort). Something as simple as `:hand: I am reviewing this` will suffice.

**Reviewer instructions**

- Please work through the checklist at the start of this issue.
- If you need any further guidance/clarification take a look at the reviewer guidelines here http://joss.theoj.org/about#reviewer_guidelines
- Please make a publication recommendation at the end of your review

Any questions, please ask for help by commenting on this issue! 🚀

**Comment by krother** [comment]
:hand: I am reviewing this

**Comment by krother** [comment]
- References: The article "Pegasus: A framework ..." has a doi that is not listed in the paper. I did not check the other doi-less articles.
- Paper: In my opinion the first paragraph of the summary deserves improvement. For instance, I find it hard to understand that Skeleton is about *distributed* applications.

**Comment by krother** [comment]
## Recommendation:
I recommend publication after a minor review of the *statement of need* both in the README and the paper. An double-check of the doi-less references would be great.

## General checks
all checked

## Functionality
all checked

- I checked the simple example from the documentaion: it worked out of the box.
- I did not check the Pegasus / Swift examples.

## Documentation

- Statement of need: needs improvement. I did not find a description of the target audience. I do not fully understand what the program is useful for. 
- Installation instructions: accurate, although the installation is not automatic. I think a plain text format for the documentation would be much easier to access than the .pptx (and would look better on LibreOffice, too). Copy-pasting from pptx file did not work for this command:

    bin/aimes-skeleton-generate -i src/aimes/skeleton/sample-input/bag.input -m shell

- Example Usage: OK
- Functionality documentation: Yes, I find the manual clearly written and detailed enough. The figures are very illustrative and help understanding. There are some minor English grammar issues but they don't impede readability.
- Automated tests: No automated tests, but a precise list of manual steps and an accurate description of an use case and the expected result.
- Community guidelines: There is an invitation link to the mailing list, but not more. I am reluctant to check point 1) "guidelines for third parties wishing to contribute"

## Software paper:
- Authors: OK
- Statement of need: In my opinion the first paragraph of the summary deserves improvement. For instance, I find it hard to understand that Skeleton is about distributed applications.
- References: The article *"Pegasus: A framework ..."* has a doi that is not listed in the paper. I did not check the other doi-less articles.

**Comment by danielskatz** [comment]
Thanks @krother.

I've added the DOIs for the 3 papers that have DOIs now and will work on the other issues.

**Comment by danielskatz** [comment]
regarding the statement of need, and your comment about distributed applications, I've changed the first paragraph to try to make the need/use of the software more clear, and changed the second paragraph to explain more about what types of applications can be built.  Please let me know if this seems ok, or if you think more changes here are needed.

**Comment by krother** [comment]
Yes it's definitely better now. I am not entirely sure what to do with the word 'application' (I think mobile app, web app, Python app.. ). I understand for someone developing distributed systems the term translates to 'the program that does the work'. Please correct me if I'm on an entirely wrong track.

**Comment by danielskatz** [comment]
I've now updated the community guidelines, and we will think about your 'application' comment a bit more, then do something tbd.

**Comment by danielskatz** [comment]
@krother : I've updated the description of the applications to address your point, and also made the description of the users/need a little more clear, I hope, in both the README and the paper.

Please let us know if this suffices, or if you think it should be further improved.

**Comment by danielskatz** [comment]
@arfon As we make changes in paper.md and the bib entries, how does the compiled paper get updated?  Or does it not until the review process ends?  It might be nice to have some way that the original submitter could recompile, or a hook that recompiles on changes to specific files in the repo once a submission to JOSS has been made.  If this makes sense, perhaps it should be in issue in the JOSS repo, rather than a comment here?

**Comment by krother** [comment]
Great, I think the text is sufficiently clear now. I'm ticking all the boxes, let's see what happens next.

I have no idea about how to update the paper. I think the JOSS repo is the right place for that.

**Comment by whedon** [comment]
Compiled paper PDF: [paper.pdf](https://github.com/openjournals/joss-reviews/files/260319/paper.pdf)

**Comment by arfon** [comment]
> @arfon As we make changes in paper.md and the bib entries, how does the compiled paper get updated? Or does it not until the review process ends? It might be nice to have some way that the original submitter could recompile, or a hook that recompiles on changes to specific files in the repo once a submission to JOSS has been made. If this makes sense, perhaps it should be in issue in the JOSS repo, rather than a comment here?

@danielskatz - yeah, please file an issue in the main JOSS repo (https://github.com/openjournals/joss). Ultimately I would like this process to be automated.

**Comment by danielskatz** [comment]
Please update the paper again, as there have been a couple of typos fixes.  Otherwise, I think this is done. - thanks @krother

**Comment by whedon** [comment]
Compiled paper PDF: [paper.pdf](https://github.com/openjournals/joss-reviews/files/260378/paper.pdf)

**Comment by arfon** [comment]
@krother - many thanks for the extremely rapid review :zap:

**Comment by arfon** [comment]
This is now accepted and live on JOSS!

- http://joss.theoj.org/papers/10.21105/joss.00017
- http://dx.doi.org/10.21105/joss.00017

**Comment by hausen** [comment]
Some chars (the curly quotes and the dashes) are garbled in the reference section.

**Comment by arfon** [comment]
> Some chars (the curly quotes and the dashes) are garbled in the reference section (web version only, the PDF is ok)

Yeah, not sure what's going on there. Will investigate.

**Comment by arfon** [comment]
> Some chars (the curly quotes and the dashes) are garbled in the reference section (web version only, the PDF is ok)

@hausen this should be fixed now. Can you confirm things are looking good for you now?

**Comment by hausen** [comment]
It's ok now.

**Comment by krother** [comment]
Hi Arfon,

this is looking good. Great job starting JOSS. I am very happy about the process!

Best regards,
Kristian

----

Dr. Kristian Rother
(+49) 0176 3052 4691
www.academis.eu

Create Better Science!

**Comment by danielskatz** [comment]
Thanks everyone - this was a great process to experience as an author!

**Comment by Spencerx** [comment]
Some clarity through guidelines should avoid doubt when another review has
to be done sooner or later to avoid any doubt or confusion.
